The ``101-lazy_matrix_mul`` module
==================================

Using ``lazy_matrix_mul``
-------------------------

Importing function from the module:

    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

Multiplying two matrices

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]])
    [[7, 10], [15, 22]]

Multiplying two matrices 2

    >>> lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]])
    [[13, 16]]

Multiplying two matrices 3

    >>> list1 = [[1, 4, 7], [2, 5, 8], [3, 6, 9]]
    >>> list2 = [[1, -1, 2], [2, -1, 2], [3, -3, 0]]
    >>> lazy_matrix_mul(list1, list2)
    [[30, -26, 10], [36, -31, 14], [42, -36, 18]]

Multiplying two matrices 4

    >>> lazy_matrix_mul([[1, 5, 7]], [[3], [5], [8]])
    [[84]]

Passing None as a matrix

    >>> lazy_matrix_mul(None, None)
    Traceback (most recent call last):
          ...
    ValueError: shapes (1,1) and (1,1) not aligned: 1 (dim 1) != 1 (dim 0)

Passing None as a matrix

    >>> lazy_matrix_mul([[7, 12]], None)
    Traceback (most recent call last):
          ...
    ValueError: shapes (1,2) and (1,1) not aligned: 2 (dim 1) != 1 (dim 0)

Passing a tuple as a matrix

    >>> lazy_matrix_mul((1, 3, 5), [[1.3], [5.2], [7.7]])
    Traceback (most recent call last):
          ...
    ValueError: shapes (3,) and (3,1) not aligned: 3 (dim 0) != 3 (dim 0)

Passing a string as a matrix

    >>> lazy_matrix_mul([[3, 3], [4, 4]], "Gintama")
    Traceback (most recent call last):
          ...
    ValueError: Scalar operands are not allowed, use '*' instead

Passing a list of tuples

    >>> lazy_matrix_mul([(1, 5, 10)], [(1), (5)])
    Traceback (most recent call last):
          ...
    ValueError: shapes (1,3) and (2,) not aligned: 3 (dim 1) != 2 (dim 0)

Passing an empty matrix

    >>> lazy_matrix_mul([], [[1]])
    Traceback (most recent call last):
          ...
    ValueError: shapes (1,1) and (1,1) not aligned: 1 (dim 1) != 1 (dim 0)

Passing an empty matrix 2

    >>> lazy_matrix_mul([[1, 2]], [[]])
    Traceback (most recent call last):
          ...
    ValueError: shapes (1,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

Passing a matrix with a list of strings

    >>> lazy_matrix_mul([[1, '1', 1], ['2', 2, '2']], [[3], [3]])
    Traceback (most recent call last):
          ...
    ValueError: invalid number of columns in each row

Passing a matrix with a list of strings 2

    >>> lazy_matrix_mul([[2, 4, 6], [3, 6, 9]], [[1], ['2'], [3]])
    Traceback (most recent call last):
          ...
    ValueError: invalid number of columns in each row

Passing a matrix with an empty list

    >>> lazy_matrix_mul([[12, 12, 12], [], [14, 14, 14]], [[2], [3], [8]])
    Traceback (most recent call last):
          ...
    ValueError: shapes (3,3) and (3,1) not aligned: 3 (dim 1) != 3 (dim 0)

Passing a matrix which its rows have different size

    >>> lazy_matrix_mul([[2, 2, 2], [4, 4, 4]], [[2], [3, 3]])
    Traceback (most recent call last):
          ...
    ValueError: all the input array dimensions for the concatenation axis must match exactly

Passing two matrix that can't be multiplied

    >>> lazy_matrix_mul([[5, 2, 3], [7, 13, 2], [1, 0, 3]], [[2, 3], [6, 9]])
    Traceback (most recent call last):
          ...
    ValueError: shapes (3,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

Missing one argument

    >>> lazy_matrix_mul([[1, 2]])
    Traceback (most recent call last):
          ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

Missing arguments

    >>> lazy_matrix_mul()
    Traceback (most recent call last):
          ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
